Roadmap 2025 (#37)
ethereum/js-team-organization: roadmap/EF2025.md

# EF JavaScript Team Roadmap 2025

This document reflects our initial planning for the respective year and represents
the roadmap we have submitted internally within the EF.

## Overview

### The EF JavaScript Team

The EF JavaScript team consists of 7 team members and is currently led by [Holger Drewes](https://github.com/holgerd77).
It is historically grown out of the - initially - community driven work on an Ethereum L1 protocol stack implemented in
`JavaScript` for Node.js and browser usage, called [EthereumJS](https://github.com/ethereumjs/ethereumjs-monorepo).

> Over the years this stack matured a lot and with it's modular nature (all 15 libraries can be used in isolation) it now

Reviewer: ```suggestion
Over the years this stack matured a lot and with its modular nature (all 15 libraries can be used in isolation) it now
```

> serves as a robust platform for both productive usage of activated L1 Ethereum Mainnet technolgies and applied EIP research and

Reviewer: ```suggestion
serves as a robust platform for both production usage of activated L1 Ethereum Mainnet technologies and applied EIP research and
```

[stateless Verkle implementation](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/statemanager#statelessverklestatemanager-experimental) ready for some time which we use to join [verkle testnets](https://github.com/ethpandaops/verkle-devnets) to help to test and evolve the verkle specification (see e.g. [here](https://notes.ethereum.org/@ethpandaops/verkle-gen-devnet-6)
for an overview).

Throughout the year we have done additional work to make the interfaces within the libraries more tree-agnostic (so: that it does not matter
if a Merkle or Verkle tree is used) and brought our own [verkle tree implementation](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/verkle) to the point where we now can do stateful verkle execution.

Together with the new test generation capabilities for the [cross client tests](https://github.com/ethereum/execution-spec-tests/releases/tag/verkle%40v0.0.5) in `execution-spec-tests` we have gained (see "Protocol Testing" track) we have now everything together to support Verkle development on a
broader level.

### 2025 Deliverables

- Continued early-on spec implementations and feedback

> - Help setup and run stateless verkle testnets

Reviewer: ```suggestion
- Help setup and run stateless verkle testnets
- Provide a JavaScript verkle trie implementation, along with proof verification and generation
- Allow stateful verkle block execution
```

- Active contributions to work on verkle test coverage and fill tests

## Stateless Verkle Client & Modularization

Track responsible: [Amir Ghorbanian](https://github.com/scorbajio)

Stateless will make (parts of) a browser based/used client more viable. We have all the building blocks together (Verkle, VM, client) to serve software
components for this use case. 2025 will be the year where we can bundle all existing pieces together respectively refactor in a way so that browser
applications will be able to integrate lightweight verkle block verification into their code base. This will allow developers to build their dApps in a
more trustworthy way by rely on less 3rd party trust assumptions.

### 2025 Deliverables

- Modularization of the EthereumJS client code
- App -> Library: reusage/integration into existing code bases
- Strong client dependency reduction to prepare for production use cases
- Verkle-tailored stateless client version of the client
- Work on outer UX experience

## Portal Network & EIP-4444

Track resonsible: [Scotty Poi](https://github.com/ScottyPoi)

> [Portal network](https://github.com/ethereum/portal-network-specs) evolved throughout 2024 and is now at a point where the specs matured for early-developed networks like the history network and is now considered for production usage e.g. in the form of being a source for providing chain history in an [EIP-4444](https://eips.ethereum.org/EIPS/eip-4444) context. The JavaScript team is developing its own [Ultralight](https://github.com/ethereumjs/ultralight) portal client and contributed to Portal development with an early on state network implementation (see e.g. PRs [#539](https://github.com/ethereumjs/ultralight/pull/539) and [#649](https://github.com/ethereumjs/ultralight/pull/649)) as well as providing robust bootnodes for setting up test networks, among other things.

Reviewer 2: I would be hesitant to say our bootnodes are "robust" unless @ScottyPoi signs off on it.  :grimacing:

Author: Have now fully used Scotty's version

- Own selected browser-based EIP community showcases

## Protocol Testing

Track responsible: [Jochem Brouwer](https://github.com/jochem-brouwer)

Building upon our active EIP and HF developement and our previous cross-client testing experiences we have now finished the integration of our VM into the new [execution-spec-tests](https://github.com/ethereum/execution-spec-tests/) cross-client testing environment in late 2024 (see PR [#3603](https://github.com/ethereumjs/ethereumjs-monorepo/pull/3603)).

We have build up some first test case creation experience afterward, culminating in the filling of the [pectra-devnet-4](https://github.com/ethereum/execution-spec-tests/releases/tag/pectra-devnet-4%40v1.0.0) tests with the EthereumJS based integration.

In 2025 we plan to build upon this experiences and become a regular test case contributor.

### 2025 Deliverables

- Continued test filling support
- Establish in "team DNA" to create tests cases on a regular basis
- Active test creation work in selected areas (EOF, 7702,...)
- More structured approach on test coverage / edge case detection

## Research Prototyping

Track responsible: [Gajinder Singh](https://github.com/g11tech)

> For the same reasons of our stack being quite modular and extensive, the EthereumJS libraries are well suited to do early-on research prototyping. In 2024 we have leveraged this by realizing an extensive prototype implementing a research snapshot around the work on a new SSZ-based transaction format ([EIP-6404](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-6404.md)) as well as a corresponding signature scheme ([EIP-6493](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-6493.md)) + additional SSZ-ification side EIPs.

Reviewer 3: ```suggestion
For the same reasons of our stack being quite modular and extensive, the EthereumJS libraries are well suited to do early-on research prototyping. In 2024 we have leveraged this by realizing an extensive prototype implementing a research snapshot around the work on:
- a new SSZ-based transaction format ([EIP-6404](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-6404.md)) as well as a corresponding signature scheme ([EIP-6493](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-6493.md))
- additional SSZ-ification side EIPs (like [EIP-6466](https://eips.ethereum.org/EIPS/eip-6466)
- further enabling proofing capabilities (like [EIP-7792](https://eips.ethereum.org/EIPS/eip-7792), [EIP-7708](https://eips.ethereum.org/EIPS/eip-7708))
```
